File: backend/app/models/circuito.py
from sqlalchemy import Column, Integer, String, Float, DateTime, ForeignKey, Enum, Boolean, Text, JSON, Index
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base
import enum

class TipoCabo(str, enum.Enum):
    CU_PVC  = "CU-PVC"
    CU_XLPE = "CU-XLPE"
    AL_PVC  = "AL-PVC"
    AL_XLPE = "AL-XLPE"

class Circuito(Base):
    __tablename__ = "circuitos"

    id           = Column(Integer, primary_key=True, index=True)
    projeto_id   = Column(Integer, ForeignKey("projetos.id"), nullable=False)
    ordem        = Column(Integer, default=0)
    descricao    = Column(String(200), nullable=False)
    tensao       = Column(Float, default=380)
    potencia_kw  = Column(Float, nullable=False)
    # --- MT/AT fields (nullable) ---
    classe_tensao_kv = Column(Float, nullable=True)
    nbi_kv = Column(Float, nullable=True)
    tafi_ka = Column(Float, nullable=True)
    sequencia_operacao = Column(String(30), nullable=True)
    meio_extincao = Column(String(30), nullable=True)
    tipo_acionamento = Column(String(30), nullable=True)
    acessorios = Column(JSON, nullable=True)
    modelo = Column(String(80), nullable=True)
    norma_referencia = Column(String(80), nullable=True)
    fator_potencia = Column(Float, default=0.85)
    distancia_m  = Column(Float, nullable=False)
    tipo_cabo    = Column(Enum(TipoCabo), default=TipoCabo.CU_PVC)
    temp_ambiente = Column(Float, default=30)
    fases        = Column(Integer, default=3)
    agrupamento  = Column(Integer, default=1)

    # Dados executivos de entrada
    tag               = Column(String(80), nullable=True)
    from_barramento   = Column(String(120), nullable=True)
    to_equipamento    = Column(String(120), nullable=True)
    protection_device = Column(String(40), nullable=True)
    modo_dimensionamento = Column(String(20), default="manual")
    modo_selecao_componentes = Column(String(20), default="manual")
    disjuntor_tensao_nominal = Column(Float, nullable=True)
    disjuntor_corrente_nominal = Column(Float, nullable=True)
    disjuntor_icu = Column(Float, nullable=True)
    disjuntor_curva = Column(String(30), nullable=True)
    disjuntor_fabricante = Column(String(80), nullable=True)
    corrente_ac_dc    = Column(String(10), default="AC")
    potencia_kva      = Column(Float, nullable=True)
    usar_kva_informado = Column(Boolean, default=False)
    fator_demanda     = Column(Float, default=1.0)
    fator_eficiencia  = Column(Float, default=1.0)
    metodo_instalacao = Column(String(30), default="TRAY")
    formacao          = Column(Integer, default=1)
    comprimento_real  = Column(Float, nullable=True)
    queda_tensao_alimentador = Column(Float, default=0.0)

    # Resultados calculados
    corrente_nominal = Column(Float, nullable=True)
    secao_mm2        = Column(Float, nullable=True)
    disjuntor_a      = Column(Float, nullable=True)
    queda_tensao_pct = Column(Float, nullable=True)
    secao_pe_mm2     = Column(Float, nullable=True)
    ampacidade       = Column(Float, nullable=True)
    status           = Column(String(10), nullable=True)

    # Resultados executivos calculados
    corrente_projeto       = Column(Float, nullable=True)
    corrente_corrigida     = Column(Float, nullable=True)
    fator_k1               = Column(Float, nullable=True)
    fator_k2               = Column(Float, nullable=True)
    fator_k3               = Column(Float, nullable=True)
    corrente_condutor      = Column(Float, nullable=True)
    tensao_fase            = Column(Float, nullable=True)
    isc_local              = Column(Float, nullable=True)
    isc_cabo               = Column(Float, nullable=True)
    tempo_atuacao          = Column(Float, nullable=True)
    secao_joule            = Column(Float, nullable=True)
    impedancia_rdc         = Column(Float, nullable=True)
    impedancia_rac         = Column(Float, nullable=True)
    impedancia_xl          = Column(Float, nullable=True)
    queda_tensao_max       = Column(Float, nullable=True)
    queda_tensao_acumulada = Column(Float, nullable=True)
    tipo_cabo_comercial    = Column(String(80), nullable=True)
    nota_tecnica           = Column(Text, nullable=True)
    revisao                = Column(String(10), nullable=True)
    status_final           = Column(String(30), nullable=True)
    protecao_status        = Column(String(30), nullable=True)
    protecao_nota          = Column(Text, nullable=True)
    validacao_status       = Column(String(30), nullable=True)
    validacao_mensagem     = Column(Text, nullable=True)
    validacao_detalhes     = Column(Text, nullable=True)
    cabo_sugerido_secao    = Column(Float, nullable=True)
    cabo_sugerido_tipo_comercial = Column(String(80), nullable=True)
    cabo_sugerido_ampacidade = Column(Float, nullable=True)
    disjuntor_sugerido_in  = Column(Float, nullable=True)
    disjuntor_sugerido_icu = Column(Float, nullable=True)
    disjuntor_sugerido_curva = Column(String(30), nullable=True)
    selecao_componentes_status = Column(String(30), nullable=True)
    selecao_componentes_justificativa = Column(Text, nullable=True)

    criado_em     = Column(DateTime(timezone=True), server_default=func.now())
    atualizado_em = Column(DateTime(timezone=True), onupdate=func.now())

    projeto = relationship("Projeto", back_populates="circuitos")

    # ── Performance indexes ─────────────────────────────────────────────────
    # ix_circuitos_projeto_ordem: composite covering index for the single most
    # common query: WHERE projeto_id = ? ORDER BY ordem.
    # With 100k circuits across all projects, this eliminates full-table scans.
    __table_args__ = (
        Index("ix_circuitos_projeto_ordem", "projeto_id", "ordem"),
    )

File: backend-node/src/models/Circuito.js
'use strict'

const mongoose = require('mongoose')

/**
 * Circuito — espelho exato do model Python (circuito.py)
 * 52 campos mapeados 1:1 com tipos Mongoose equivalentes
 */
const CircuitoSchema = new mongoose.Schema(
  {
    projetoId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Projeto',
      required: true,
      index: true,
      immutable: true,
    },
    ordem:    { type: Number, default: 0, min: 0 },
    descricao: { type: String, required: true, trim: true, maxlength: 200 },

    // ── Dados elétricos de entrada ─────────────────────────────────────────
    tensao:       { type: Number, default: 380, min: 1, max: 2000000 },
    tensao_unidade: { type: String, enum: ['V', 'kV'], default: 'V' },
    tipo_sistema_tensao: { type: String, enum: ['AC', 'DC'], default: 'AC' },
    referencia_tensao_dc: { type: String, default: null, trim: true, maxlength: 80 },
    contexto_aplicacao: { type: String, default: null, trim: true, maxlength: 120 },
    potencia_kw:  { type: Number, required: true, min: 0, max: 1000000 },
    potencia_kva: { type: Number, default: null, min: 0, max: 1000000 },
    fator_potencia:       { type: Number, default: 0.85, min: 0.01, max: 1 },
    distancia_m:          { type: Number, required: true, min: 0, max: 100000 },
    tipo_cabo: {
      type: String,
      enum: ['CU-PVC', 'CU-XLPE', 'AL-PVC', 'AL-XLPE'],
      default: 'CU-PVC',
    },
    temp_ambiente:  { type: Number, default: 30, min: -50, max: 120 },
    fases:          { type: Number, default: 3, enum: [1, 2, 3] },
    agrupamento:    { type: Number, default: 1, min: 1, max: 200 },
    corrente_ac_dc: { type: String, enum: ['AC', 'DC'], default: 'AC' },
    fator_demanda:           { type: Number, default: 1.0, min: 0, max: 1 },
    fator_eficiencia:        { type: Number, default: 1.0, min: 0.01, max: 1 },
    metodo_instalacao:       { type: String, default: 'TRAY', trim: true, maxlength: 80 },
    formacao:                { type: Number, default: 1, min: 1, max: 20 },
    comprimento_real:        { type: Number, default: null, min: 0 },
    queda_tensao_alimentador:{ type: Number, default: 0.0, min: 0 },
    isc_local:               { type: Number, default: null, min: 0 },
    usar_kva_informado:      { type: Boolean, default: false },
    configuracao_eletrica:   { type: String, default: null, trim: true, maxlength: 200 },
    referencia_tensao:       { type: String, default: null, trim: true, maxlength: 200 },
    modo_entrada:            { type: String, default: null, trim: true, maxlength: 200 },
    base_potencia:           { type: String, default: null, trim: true, maxlength: 200 },
    corrente_informada:      { type: Number, default: null, min: 0 },
    aplicacao_circuito:      { type: String, default: null, trim: true, maxlength: 200 },
    secao_minima_aplicacao:  { type: Number, default: null, min: 0 },
    secao_minima_mecanica:   { type: Number, default: null, min: 0 },
    queda_tensao_limite:     { type: Number, default: null, min: 0 },

    // ── Dados executivos de entrada ────────────────────────────────────────
    tag:              { type: String, default: null, trim: true, maxlength: 200 },
    from_barramento:  { type: String, default: null, trim: true, maxlength: 200 },
    to_equipamento:   { type: String, default: null, trim: true, maxlength: 200 },
    protection_device:{ type: String, default: null, trim: true, maxlength: 200 },
    modo_dimensionamento:       { type: String, default: 'manual', trim: true, maxlength: 80 },
    modo_selecao_componentes:   { type: String, default: 'manual', trim: true, maxlength: 80 },
    disjuntor_tensao_nominal:   { type: Number, default: null, min: 0 },
    disjuntor_corrente_nominal: { type: Number, default: null, min: 0 },
    disjuntor_icu:              { type: Number, default: null, min: 0 },
    disjuntor_curva:            { type: String, default: null, trim: true, maxlength: 80 },
    disjuntor_fabricante:       { type: String, default: null, trim: true, maxlength: 200 },
    protecao_curva_fonte:       { type: String, default: null, trim: true, maxlength: 500 },
    protecao_curva_pontos:      { type: [mongoose.Schema.Types.Mixed], default: [] },

    // ── Campos MT/AT ───────────────────────────────────────────────────────
    classe_tensao_kv:   { type: Number, default: null, min: 0 },
    nbi_kv:             { type: Number, default: null, min: 0 },
    tafi_ka:            { type: Number, default: null, min: 0 },
    sequencia_operacao: { type: String, default: null, trim: true, maxlength: 200 },
    meio_extincao:      { type: String, default: null, trim: true, maxlength: 200 },
    tipo_acionamento:   { type: String, default: null, trim: true, maxlength: 200 },
    acessorios:         { type: mongoose.Schema.Types.Mixed, default: null },
    modelo:             { type: String, default: null, trim: true, maxlength: 200 },
    norma_referencia:   { type: String, default: null, trim: true, maxlength: 200 },

    // ── Resultados calculados (preenchidos pelo FastAPI) ───────────────────
    corrente_nominal:  { type: Number, default: null },
    secao_mm2:         { type: Number, default: null },
    disjuntor_a:       { type: Number, default: null },
    queda_tensao_pct:  { type: Number, default: null },
    secao_pe_mm2:      { type: Number, default: null },
    ampacidade:        { type: Number, default: null },
    status:            { type: String, default: null },

    corrente_projeto:   { type: Number, default: null },
    corrente_corrigida: { type: Number, default: null },
    fator_k1:           { type: Number, default: null },
    fator_k2:           { type: Number, default: null },
    fator_k3:           { type: Number, default: null },
    corrente_condutor:  { type: Number, default: null },
    ampacidade_corrigida_total: { type: Number, default: null },
    tensao_fase:        { type: Number, default: null },
    isc_cabo:           { type: Number, default: null },
    tempo_atuacao:      { type: Number, default: null },
    secao_joule:        { type: Number, default: null },
    impedancia_rdc:     { type: Number, default: null },
    impedancia_rac:     { type: Number, default: null },
    impedancia_xl:      { type: Number, default: null },
    queda_tensao_max:       { type: Number, default: null },
    queda_tensao_acumulada: { type: Number, default: null },
    tipo_cabo_comercial:    { type: String, default: null },
    nota_tecnica:           { type: String, default: null },
    revisao:                { type: String, default: null },
    status_final:           { type: String, default: null },
    protecao_status:        { type: String, default: null },
    protecao_nota:          { type: String, default: null },
    protecao_avaliacao_status: { type: String, default: null },
    protecao_verificacoes:  { type: mongoose.Schema.Types.Mixed, default: null },
    validacao_status:       { type: String, default: null },
    validacao_mensagem:     { type: String, default: null },
    validacao_detalhes:     { type: mongoose.Schema.Types.Mixed, default: null },

    cabo_sugerido_secao:          { type: Number, default: null },
    cabo_sugerido_tipo_comercial: { type: String, default: null },
    cabo_sugerido_ampacidade:     { type: Number, default: null },
    disjuntor_sugerido_in:        { type: Number, default: null },
    disjuntor_sugerido_icu:       { type: Number, default: null },
    disjuntor_sugerido_curva:     { type: String, default: null },
    selecao_componentes_status:         { type: String, default: null },
    selecao_componentes_justificativa:  { type: String, default: null },

    // Snapshot auditável aditivo; campos planos legados continuam disponíveis.
    resultado:          { type: mongoose.Schema.Types.Mixed, default: null },
    criterios:          { type: mongoose.Schema.Types.Mixed, default: null },
    decisao:            { type: mongoose.Schema.Types.Mixed, default: null },
    memorial:           { type: mongoose.Schema.Types.Mixed, default: null },
    alertas:            { type: [mongoose.Schema.Types.Mixed], default: [] },
    premissas:          { type: [mongoose.Schema.Types.Mixed], default: [] },
    limitacoes:         { type: [String], default: [] },
    metadados_calculo:  { type: mongoose.Schema.Types.Mixed, default: null },
  },
  {
    timestamps: { createdAt: 'criado_em', updatedAt: 'atualizado_em' },
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
)

// Cobrindo index: WHERE projetoId = ? ORDER BY ordem  (idêntico ao Python)
CircuitoSchema.index({ projetoId: 1, ordem: 1 })
CircuitoSchema.index({ projetoId: 1, tag: 1 }, { sparse: true })
CircuitoSchema.index({ projetoId: 1, status_final: 1 })

module.exports = mongoose.model('Circuito', CircuitoSchema)
